pkg/cafs/cafs.go

```diff
@@ -102,28 +64,56 @@ type Fs interface {
 	GetAddressingScheme() string
 }
 
-// New creates a new file system operations instance for a repository
-func New(opts ...Option) (Fs, error) {
+var _ Fs = &defaultFs{}
 
-	f := &defaultFs{
+func defaultsForFs() *defaultFs {
+	return &defaultFs{
 		store:                       cafsStore{backend: localfs.New(nil)},
-		leafSize:                    uint32(5 * units.MiB),
+		leafSize:                    DefaultLeafSize,
 		concurrentFlushes:           10,
 		readerConcurrentChunkWrites: 3,
 		deduplicationScheme:         DeduplicationBlake,
+		keysCacheSize:               DefaultKeysCacheSize,
+		lruSize:                     DefaultCacheSize,
+		l:                           dlogger.MustGetLogger("info"),
+		withVerifyHash:              true,
+		withPrefetch:                0, // prefetching disabled by default
 	}
-	f.leafPool = newLeafFreelist()
-	f.lru, _ = lru.NewWithEvict(10, func(lruKey interface{}, lruVal interface{}) {
-		lbuf := lruVal.(*leafBuffer)
-		f.leafPool.put(lbuf)
-	})
-	f.lru, _ = lru.New(10)
+}
+
+// New creates a new instance of a content-addressable file system
+func New(opts ...Option) (Fs, error) {
+	f := defaultsForFs()
 	for _, apply := range opts {
 		apply(f)
 	}
+
 	if f.leafSize > MaxLeafSize {
 		return nil, fmt.Errorf("%v exceeds maximum cafs leaf size %v", f.leafSize, MaxLeafSize)
 	}
+	if f.leafSize < KeySize {
+		return nil, fmt.Errorf("%v is smaller than the key size %v", f.leafSize, KeySize)
+	}
+
+	const buffersForparallelReaders = 3
+	cacheBuffers := BytesToBuffers(f.lruSize, f.leafSize)
+	f.leafPool = newLeafFreelist(f.leafSize, cacheBuffers+buffersForparallelReaders)
+
+	var err error
+	f.lru, err = lru.NewWithEvict(cacheBuffers, func(_ interface{}, lruVal interface{}) {
+		f.leafPool.Release(lruVal.(LeafBuffer)) // relinquish buffers to the freelist
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	f.keysCache, err = lru.New(f.keysCacheSize)
```

Comment:
let's either attach a freelist here or make a note to profile memory usage:  it could be that the `lru` causes GC to lose track of `alloc`s in all cases, or it could be that what's going on with the leaf buffers is unique to that cache.

Reply:
The default reader takes  a nil instance of the freelist, which means it is just a wrapper around the gc, without any specific eviction mechanism. I am using this for testing (chunkReader testing has been augmented to cover several combinations of the options at hand). The default for the fs is to use a freelist with a high watermark tuned to the LRU size plus some read concurrency setting (which is grossly determined as for now)

Reply:
follow-up note: open discussion thread

Reply:
making a mental note to profile memory usage here in case it is an issue in order to resolve the thread